File: internal/router/responses.go
// /v1/responses — OpenAI Responses-API entry point. Codex subscription is
// the only provider that's "Responses-native" today; this handler errors out
// for everything else.
package router

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/amaozhao/lazirouter/internal/combo"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/httpx"
	"github.com/amaozhao/lazirouter/internal/logger"
	"github.com/amaozhao/lazirouter/internal/picker"
	"github.com/amaozhao/lazirouter/internal/provider"
	"github.com/amaozhao/lazirouter/internal/quota"
	"github.com/amaozhao/lazirouter/internal/tenantctx"
	"github.com/amaozhao/lazirouter/internal/usage"
	"github.com/google/uuid"
)

func (d *Deps) handleResponses(w http.ResponseWriter, r *http.Request) {
	started := time.Now()
	reqID := uuid.NewString()

	apiKey := httpx.BearerOrXAPIKey(r)
	if apiKey == "" {
		httpx.WriteError(w, errs.Auth("Missing Authorization or x-api-key header"))
		return
	}
	ctx, err := tenantctx.ResolveAPIKey(r.Context(), apiKey)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := tenantctx.EnforceTenantActive(ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := quota.Enforce(r.Context(), ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}

	body, err := httpx.ReadJSONMap(r)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	model, _ := body["model"].(string)
	if model == "" {
		httpx.WriteError(w, errs.Validation("Missing field: model"))
		return
	}
	if _, hasInput := body["input"]; !hasInput {
		if _, hasMsgs := body["messages"].([]any); !hasMsgs {
			httpx.WriteError(w, errs.Validation("Missing field: input[] or messages[]"))
			return
		}
	}
	// Default to streaming because the Codex backend doesn't support non-stream.
	stream := true
	if v, ok := body["stream"]; ok && v == false {
		stream = false
	}

	attempts, err := combo.Resolve(r.Context(), ctx.TenantID, model)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}

	var errsList []map[string]any
	for _, attempt := range attempts {
		if attempt.Provider != "codex" {
			errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": "provider " + attempt.Provider + " does not expose a Responses API path"})
			continue
		}
		account, err := picker.Pick(r.Context(), d.Cfg.MasterKey, ctx.TenantID, attempt.Provider)
		if err != nil {
			var ae *errs.AppError
			if errors.As(err, &ae) && ae.Code == "no_account_available" {
				errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": ae.Message})
				continue
			}
			httpx.WriteError(w, err)
			return
		}
		body["model"] = attempt.UpstreamModel
		sc := provider.SessionContext{TenantID: ctx.TenantID, ConnectionID: account.ConnectionID}
		ueBase := &usage.Event{
			TenantID: ctx.TenantID, APIKeyID: ctx.APIKeyID, ConnectionID: account.ConnectionID,
			Provider: attempt.Provider, Model: model, UpstreamModel: attempt.UpstreamModel,
			RequestID: reqID, Meta: map[string]any{"transport": "/v1/responses"},
		}
		var callErr error
		if stream {
			callErr = d.streamResponses(r.Context(), w, account, body, sc, ueBase, started)
		} else {
			callErr = d.nonStreamResponses(r.Context(), w, account, body, sc, ueBase, started)
		}
		if callErr == nil {
			return
		}
		var ae *errs.AppError
		status := 0
		if errors.As(callErr, &ae) {
			if sc, ok := ae.Meta["status_code"].(int); ok {
				status = sc
			}
		}
		retriable := status == 429 || (status >= 500 && status < 600) || !errors.As(callErr, &ae)
		if retriable && status > 0 {
			_, _ = picker.MarkCooldown(r.Context(), ctx.TenantID, attempt.Provider, account.ConnectionID)
		}
		errsList = append(errsList, map[string]any{"step": attempt.Step, "provider": attempt.Provider, "status": status, "message": callErr.Error()})
		if !retriable {
			httpx.WriteError(w, callErr)
			return
		}
	}

	httpx.WriteError(w, errs.Upstream("All upstream attempts failed", map[string]any{"attempts": errsList}))
}

func (d *Deps) nonStreamResponses(ctx context.Context, w http.ResponseWriter, account *picker.Account, body map[string]any, sc provider.SessionContext, ue *usage.Event, started time.Time) error {
	raw, err := d.CodexSub.ResponsesCall(ctx, account.Credentials, account.Metadata, body, sc)
	if err != nil {
		_ = usage.Record(ctx, errEvent(ue, started, err))
		return err
	}
	prompt, completion := extractResponsesUsage(raw)
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	// Codex backend always streams; we forward the SSE text under an
	// event-stream content type so clients can parse frames the same way.
	w.Header().Set("content-type", "text/event-stream")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(raw)
	return nil
}

func (d *Deps) streamResponses(ctx context.Context, w http.ResponseWriter, account *picker.Account, body map[string]any, sc provider.SessionContext, ue *usage.Event, started time.Time) error {
	frames, errc := d.CodexSub.StreamResponses(ctx, account.Credentials, account.Metadata, body, sc)
	var prompt, completion int64
	headersSent := false
	for {
		select {
		case frame, ok := <-frames:
			if !ok {
				goto done
			}
			if !headersSent {
				w.Header().Set("content-type", "text/event-stream")
				w.Header().Set("cache-control", "no-cache, no-transform")
				w.Header().Set("connection", "keep-alive")
				w.WriteHeader(http.StatusOK)
				headersSent = true
			}
			_, _ = w.Write(frame)
			if f, ok := w.(http.Flusher); ok {
				f.Flush()
			}
			if p, c := extractResponsesUsage(frame); p > 0 {
				prompt, completion = p, c
			}
		case err := <-errc:
			if !headersSent {
				_ = usage.Record(ctx, errEvent(ue, started, err))
				return err
			}
			logger.Warn("codex stream error after headers sent", "err", err)
			goto done
		case <-ctx.Done():
			return nil
		}
	}
done:
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	return nil
}

// extractResponsesUsage parses the last `event: response.completed` frame from
// a chunk of SSE text. Image_gen / web_search usage objects are nested inside
// other events and zero-valued, so we must target response.completed precisely.
func extractResponsesUsage(buf []byte) (int64, int64) {
	lines := strings.Split(string(buf), "\n")
	var event, lastData string
	for _, ln := range lines {
		if strings.HasPrefix(ln, "event: ") {
			event = strings.TrimSpace(ln[7:])
		} else if strings.HasPrefix(ln, "data: ") && event == "response.completed" {
			lastData = ln[6:]
		} else if ln == "" {
			event = ""
		}
	}
	if lastData == "" {
		return 0, 0
	}
	var obj map[string]any
	if err := json.Unmarshal([]byte(lastData), &obj); err != nil {
		return 0, 0
	}
	resp, _ := obj["response"].(map[string]any)
	usage, _ := resp["usage"].(map[string]any)
	return toI64(usage["input_tokens"]), toI64(usage["output_tokens"])
}

// scanAnthropicUsage pulls input/output tokens from Anthropic stream frames
// (event: message_delta or message_start with usage in the data payload).
func scanAnthropicUsage(buf []byte) (int64, int64) {
	lines := strings.Split(string(buf), "\n")
	for _, ln := range lines {
		if !strings.HasPrefix(ln, "data: ") {
			continue
		}
		var obj map[string]any
		if err := json.Unmarshal([]byte(ln[6:]), &obj); err != nil {
			continue
		}
		// usage may be at top level (message_start) or nested under message.usage
		// or inside message_delta.usage.
		if u, ok := obj["usage"].(map[string]any); ok {
			p := toI64(u["input_tokens"])
			c := toI64(u["output_tokens"])
			if p > 0 || c > 0 {
				return p, c
			}
		}
		if msg, ok := obj["message"].(map[string]any); ok {
			if u, ok := msg["usage"].(map[string]any); ok {
				return toI64(u["input_tokens"]), toI64(u["output_tokens"])
			}
		}
	}
	return 0, 0
}

File: router/src/routes/responses.js
// POST /v1/responses — OpenAI Responses-API entry point.
//
// Codex CLI (wire_api="responses") and any caller that speaks the
// Responses API can target this endpoint. The route mirrors the chat /
// messages pipeline (auth → rate limit → combo → pick account → upstream →
// usage event) but the upstream call is provider-native: codex subscription
// goes to https://chatgpt.com/backend-api/codex/responses with the spoof
// headers; the OpenAI Responses path (api.openai.com/v1/responses) is
// reachable via an `openai` connection.

import crypto from 'node:crypto';
import {
  ValidationError, UpstreamError, AuthError,
  NoAccountAvailableError, AppError,
} from '@lazirouter-cloud/shared';
import { enforceDailyQuota } from '@lazirouter-cloud/shared';
import { resolveApiKey, enforceTenantActive } from '../middleware/edgeAuth.js';
import { enforceRateLimit } from '../middleware/rateLimit.js';
import { pickAccount, markCooldown } from '../services/accountPicker.js';
import { recordUsage, getPricing, computeCost } from '../services/usage.js';
import { resolveAttempts } from '../services/combo.js';
import { getUpstream, isResponsesNative } from '../providers/index.js';
import { readJson } from '../lib/http.js';

export async function handleResponses(req, res) {
  const startedAt = Date.now();
  const requestId = crypto.randomUUID();

  // 1) Auth — both Bearer and x-api-key accepted to match SDK conventions
  const authHeader = req.headers['authorization'] || '';
  const bearer = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : '';
  const xApi = typeof req.headers['x-api-key'] === 'string' ? req.headers['x-api-key'] : '';
  const apiKey = bearer || xApi;
  if (!apiKey) throw new AuthError('Missing Authorization or x-api-key header');
  const ctx = await resolveApiKey(apiKey);
  enforceTenantActive(ctx);
  await enforceRateLimit(ctx);
  await enforceDailyQuota(ctx);

  const body = await readJson(req);
  if (!body.model) throw new ValidationError('Missing field: model');
  if (!body.input && !Array.isArray(body.messages)) {
    throw new ValidationError('Missing field: input[] or messages[]');
  }

  // 2) Attempts (combo / explicit provider / heuristic)
  const attempts = await resolveAttempts(ctx.tenantId, body.model);
  // Default to streaming because the Codex backend doesn't support non-stream
  // and most callers prefer it anyway.
  const isStream = body.stream !== false;

  const errors = [];
  for (let i = 0; i < attempts.length; i++) {
    const attempt = attempts[i];
    let account;
    try {
      account = await pickAccount(ctx.tenantId, attempt.provider);
    } catch (e) {
      if (e instanceof NoAccountAvailableError) {
        errors.push({ step: attempt.step, provider: attempt.provider, reason: e.message });
        continue;
      }
      throw e;
    }

    const upstream = getUpstream(attempt.provider);
    const usesResponsesNative = isResponsesNative(attempt.provider);

    if (!usesResponsesNative) {
      errors.push({ step: attempt.step, reason: `provider ${attempt.provider} does not expose a Responses API path` });
      continue;
    }
    if (!account.credentials.access_token) {
      errors.push({ step: attempt.step, reason: 'missing OAuth access_token' });
      continue;
    }

    const upstreamBody = { ...body, model: attempt.upstreamModel };
    const sessionContext = { tenantId: ctx.tenantId, connectionId: account.connectionId };

    const usageBase = {
      tenantId: ctx.tenantId,
      apiKeyId: ctx.apiKeyId,
      connectionId: account.connectionId,
      provider: attempt.provider,
      model: body.model,
      upstreamModel: attempt.upstreamModel,
      requestId,
    };

    try {
      if (isStream) {
        await handleStream({
          req, res, upstream, credentials: account.credentials, metadata: account.metadata,
          upstreamBody, sessionContext, usageBase, startedAt,
        });
      } else {
        await handleNonStream({
          res, upstream, credentials: account.credentials, metadata: account.metadata,
          upstreamBody, sessionContext, usageBase, startedAt,
        });
      }
      return;
    } catch (err) {
      const status = err instanceof UpstreamError ? err.meta?.status_code : null;
      const retriable = status === 429 || (status >= 500 && status < 600) || !(err instanceof AppError);
      if (retriable && status) await markCooldown(ctx.tenantId, attempt.provider, account.connectionId);
      await recordUsage({
        ...usageBase, status: 'error',
        latencyMs: Date.now() - startedAt,
        errorCode: status ? `upstream_${status}` : (err.code || 'unknown'),
        meta: { error: err.message, attempt: attempt.step, transport: '/v1/responses' },
      });
      errors.push({ step: attempt.step, provider: attempt.provider, status, message: err.message });
      if (res.headersSent) throw err;
      if (!retriable) throw err;
    }
  }
  const allNoAccount = errors.length > 0 && errors.every(e => !e.status);
  if (allNoAccount) throw new NoAccountAvailableError({ attempts: errors });
  throw new UpstreamError('All upstream attempts failed', { attempts: errors });
}

// Pull the authoritative {input_tokens, output_tokens} from a SSE frame.
// The Responses API emits multiple usage-shaped objects per response
// (image_gen / web_search tool usage are zero-valued and nested), so a plain
// regex on "input_tokens" picks up the wrong one. We target the
// `response.completed` event specifically and JSON-parse `data.response.usage`.
function extractUsage(frame) {
  // Frames may contain several SSE events delimited by blank lines. Find the
  // last response.completed in this chunk (typical: it is the last frame).
  const lines = frame.split('\n');
  let event = null;
  let lastData = null;
  for (const ln of lines) {
    if (ln.startsWith('event: ')) event = ln.slice(7).trim();
    else if (ln.startsWith('data: ') && event === 'response.completed') lastData = ln.slice(6);
    else if (ln === '') event = null;
  }
  if (!lastData) return null;
  try {
    const obj = JSON.parse(lastData);
    const u = obj?.response?.usage;
    if (!u) return null;
    return {
      promptTokens:     Number(u.input_tokens)  || 0,
      completionTokens: Number(u.output_tokens) || 0,
    };
  } catch { return null; }
}

async function handleNonStream({ res, upstream, credentials, metadata, upstreamBody, sessionContext, usageBase, startedAt }) {
  // Codex backend always streams; we collect the stream and re-emit JSON.
  // For now we forward the SSE text as-is to a JSON-headers response so the
  // client can pick frames out — most non-streaming Responses callers parse
  // the same way.
  let buf = '';
  let promptTokens = 0;
  let completionTokens = 0;
  for await (const frame of upstream.stream({ credentials, metadata, body: upstreamBody, sessionContext })) {
    buf += frame;
    const u = extractUsage(frame);
    if (u) { promptTokens = u.promptTokens; completionTokens = u.completionTokens; }
  }
  const pricing = await getPricing(usageBase.tenantId, usageBase.provider, usageBase.upstreamModel);
  const costMicros = computeCost(pricing, promptTokens, completionTokens);
  res.writeHead(200, { 'content-type': 'text/event-stream' });
  res.end(buf);
  await recordUsage({
    ...usageBase, status: 'ok', latencyMs: Date.now() - startedAt,
    promptTokens, completionTokens, costMicros,
    meta: { transport: '/v1/responses' },
  });
}

async function handleStream({ req, res, upstream, credentials, metadata, upstreamBody, sessionContext, usageBase, startedAt }) {
  const ac = new AbortController();
  req.on('close', () => ac.abort());

  let headersSent = false;
  let promptTokens = 0;
  let completionTokens = 0;

  for await (const frame of upstream.stream({ credentials, metadata, body: upstreamBody, signal: ac.signal, sessionContext })) {
    if (!headersSent) {
      res.writeHead(200, {
        'content-type': 'text/event-stream',
        'cache-control': 'no-cache, no-transform',
        'connection': 'keep-alive',
      });
      headersSent = true;
    }
    res.write(frame);
    const u = extractUsage(frame);
    if (u) { promptTokens = u.promptTokens; completionTokens = u.completionTokens; }
  }
  res.end();

  const pricing = await getPricing(usageBase.tenantId, usageBase.provider, usageBase.upstreamModel);
  const costMicros = computeCost(pricing, promptTokens, completionTokens);
  await recordUsage({
    ...usageBase, status: 'ok', latencyMs: Date.now() - startedAt,
    promptTokens, completionTokens, costMicros,
    meta: { transport: '/v1/responses' },
  });
}
